phoenix-core/src/main/java/org/apache/phoenix/expression/util/regex/Utils.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.expression.util.regex;
+
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.phoenix.schema.SortOrder;
+import org.apache.phoenix.schema.types.PVarchar;
+
+class Utils {
+
+    public static byte[] immutableBytesWritableToBytes(ImmutableBytesWritable ptr,
```

Comment:
The way to invert is like this:

    dataType.coerceBytes(ptr, dataType, sortOrder, SortOrder.ASC);

If the sortOrder is already ascending, it's a noop. Otherwise, it'll create a new byte array, invert only the bytes pointed to by ptr, and set ptr to point to it.